sql/src/main/java/io/crate/expression/reference/sys/shard/ShardSegments.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.expression.reference.sys.shard;
+
+import com.google.common.collect.Streams;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.inject.Singleton;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.indices.IndicesService;
+
+import java.util.Iterator;
+import java.util.stream.Stream;
+
+import static com.google.common.collect.Streams.stream;
+
+@Singleton
+public class ShardSegments implements Iterable<ShardSegment> {
```

Comment:
Keep in mind that our `Iterable`s that we use have to have stable results in case they're used in joins (and are iterated over multiple times). 

I'm not sure if this is currently the case with the stream based implementation.

Reply:
What means stable results exactly ? Preserving the same order or retrieving the same amount of items or both ?

Reply:
If I'm not mistaken, each `iterator` call on a `Iterable` must always return the exact same items as in prior `iterator` calls. To reflect changes in the underlying data a new `Iterable` can return different items.


But I'm not sure if our join algorithms meanwhile changed and are less sensitive towards changes (e.g. in the ordering). I'm also not sure if currently all `Iterable` implementations of all sys tables adhere to this. I think we should have a closer look at this (and maybe also create a test to enforce this contract)

Maybe @seut has some more insights here?

Reply:
@mfussenegger Afaik this is related to all existing system tables as well and solved by https://github.com/crate/crate/blob/master/sql/src/main/java/io/crate/execution/engine/collect/sources/SystemCollectSource.java#L110.

Reply:
But I agree that the use of streams may violate the `iterator` constrain in that way that iterable elements could change.

Reply:
Ah right, due to the copy it shouldn't be a problem.

I think I mixed up a prior issue with unstable `Iterable` that we had on non-sys-tables where such a copy operation is not in place.

Reply:
ok, great.